### ISSUE TITLE
Array scalar for fixed-length string loses length information

### ISSUE THREAD

**Issue by asmeurer** [numpy/numpy#18604]
### Reproducing code example:

```python
>>> np.array([b''], dtype='S5')[0].dtype
dtype('S')
```

The `5` information in the dtype is lost. Note that if a shape () array is used instead, this issue does not occur:

```py
>>> np.array([b''], dtype='S5')[0, ...]
array(b'', dtype='|S5')
>>> np.array([b''], dtype='S5').reshape(())
array(b'', dtype='|S5'
```

### NumPy/Python version information:

```
1.21.0.dev0+20210202221217_33dc7be 3.8.6 | packaged by conda-forge | (default, Oct  7 2020, 18:42:56)
[Clang 10.0.1 ]
```

**Comment by mattip** [comment]
There is a difference between a "0d array" and an "Array scalar" (which should just be called a numpy scalar, since it is not strictly an array and has different behaviour that only closely matches 0d arrays).
```
# A scalar
>>> x = np.array([b'abc'], dtype='S5')[0]
>>> type(x)
numpy._string
>>> x.dtype  
dtype('S3')   # the minimum needed to hold the scalar string value

# Now a 0d array
>>>  y = np.array([b''], dtype='S5')[0, ...]
>>> type(y)
numpy.ndarray
>>> y.base  # this is a view into the original array, so the dtype is unchanged
array([''], dtype='|S5')
```

But maybe you knew that and were just pointing out how scalars are weird?

**Comment by seberg** [comment]
Hmmm, I agree that it is a bug and the scalar should return `S3` here and "remember" where it came from.  I am not sure if it is possible to fix (but it might be worth checking).

The problem is that our scalars are subclasses of `bytes` and `str`.  We would have to NULL-pad the string storage of the Bytes/Unicode object, but then tell it that the actual storage is shorter.  I honestly have no idea whether that might be possible, it sounds like playing with fire.

Note that we don't have `"S"` as a real construct right now, its just a weird spelling of `S0` which NumPy special cases in many places.  ("S" should not be the same as "S0" but it is.)

**Comment by mattip** [comment]
> I agree that it is a bug and the scalar should return S3 here and "remember" where it came from

Why? Is there a spec or standard that we are holding ourselves to? I think the case is stronger for the current outcome: creating a value-based dtype (no need to carefully pad, conversion to and from python types does not need to check for excess NULL characters) or reusing the source array dtype (to be more consistent with some array, in spite of losing all other information that this scalar came from that array).

**Comment by seberg** [comment]
Yeah Matti, thinking about it a bit more, you are of course right.  The fact that scalars pretend to be 0-D arrays gave also me some intuition that this type "information loss" should not happen.  The fallacy being that there is any meaningful "information loss"!

Repeating your argument in my own words/way of thinking:

The dtype instance can represent both storage as well as "behaviour" parameters:
  * String length, just like byte order, is "storage parameter".  It does not affect behaviour of the _value_ in any situation.  So just like the byte-order, we are allowed to drop it for scalars (or in fact any array-operations).
  * A example for a "behaviour parameter" is the `unit` of a `datetime64`. Which clearly must be preserved in the NumPy scalar.  (Most DTypes do not have behaviour parameters, behaviour is usually fully defined by the `DType` class so to speak.)

Mutating a storage parameter will never modify the result of a ufunc or any normal function: the behaviour of the represented value remains intact.  Thus, we do not have much or any guarantees about preserving it.

Of course, for some programmers the metadata (storage information) may be relevant and worth preserving. But I agree that we do not and should not make it a principle.  In such programs, the metadata will have to be passed around differently where necessary.

**Comment by eric-wieser** [comment]
My opinion has always been that if you don't want dtype information loss, you shouldn't let your array decay to a scalar in the first place. Unfortunately, numpy is often very pushy in giving you a scalar even when you don't want one, which can make this approach awkard...

**Comment by asmeurer** [comment]
> Is there a spec or standard that we are holding ourselves to?

To me such a standard would be "indexing an array shouldn't change its dtype".

> My opinion has always been that if you don't want dtype information loss, you shouldn't let your array decay to a scalar in the first place. Unfortunately, numpy is often very pushy in giving you a scalar even when you don't want one, which can make this approach awkard...

Seems to me that every time I come across one of the weird behaviors of array scalars it is "expected" by NumPy developers, but if you were to ask a typical NumPy user about it it would come as a surprise to them (if they even know what array scalars are in the first place). I really think they shouldn't exist at all. 

> Of course, for some programmers the metadata (storage information) may be relevant and worth preserving. But I agree that we do not and should not make it a principle. In such programs, the metadata will have to be passed around differently where necessary.

For background, this came up in the context of h5py, which uses the S dtypes to represent the HDF5 fixed-length string data type. Specifically I was storing the empty string as the fillvalue. 

It comes as a surprise to me that NumPy loses dtype information in some cases, although that could just be because I haven't dealt very much with non-basic dtypes.

**Comment by mattip** [comment]
> I really think they shouldn't exist at all.

+1. I think we all agree that all the implications of an intermediate type between arrays and python objects were not fully appreciated at the time they were created and that people rarely actually **want** them (they just kind of "appear" when indexing returns a single value). One of the reasons I have heard for prefering them over 0d arrays is performance. Because they cannot be indexed or assigned to like a container, much of the overhead around `ndarray`s is not needed for scalars.

But here we are: they exist, and have their own rules that are now baked in to the accepted numpy ecosystem. Can we deprecate them? Maybe. It is a big project with perhaps implications that we do not fully appreciate right now, and probably cannot be done until the dtype refactor is complete.

Maybe correcting the documentation would help. The description of [scalars in the documentation](https://numpy.org/devdocs/user/basics.types.html?highlight=scalars#array-scalars) uses the `array` term, when there is nothing "array" like about them (they are not containers), It also states "The primary advantage of using array scalars is that they preserve the array type", which is wrong for flexible types.

**Comment by mattip** [comment]
xref issue gh-17744 which also expresses confusion around "array scalars", "scalar arrays" (which only appears in the documentation in very old release notes and in the swig interface page),  and "0d arrays"
